File: yt/yt/flow/examples/python/word_count/word_count_mapper.py
"""WordCountMapper: RowFunction that counts words using YSON state."""

from yt.yt.flow.library.python.companion.computation import RowFunction


# [BEGIN word_count_mapper]
class WordCountMapper(RowFunction):
    """Processes Word messages and maintains word count state."""

    def on_message(self, message, output, ctx):
        state = ctx.state("word-state", message)
        data = state.get_or_default({"word": message.payload["word"], "count": 0})
        data["count"] += 1
        state.set(data)
# [END word_count_mapper]

File: yt/yt/flow/examples/go/word_count/word_count_mapper.go
package main

import (
	"context"

	"go.ytsaurus.tech/yt/go/flow"
)

const wordStateName = "word-state"

// [BEGIN word_count_state]

type wordCountState struct {
	Word  string `yson:"word"`
	Count int64  `yson:"count"`
}

// [END word_count_state]

// [BEGIN word_count_mapper]

type wordCountMapper struct{}

var _ flow.RowFunction = (*wordCountMapper)(nil)

func (*wordCountMapper) OnMessage(
	ctx context.Context,
	rt flow.Runtime,
	msg flow.ExtendedMessage,
	out flow.OutputCollector,
) error {
	var input wordMessage
	if err := msg.ConvertTo(&input); err != nil {
		return err
	}

	state, err := flow.OpenYSONState[wordCountState](rt, wordStateName, msg)
	if err != nil {
		return err
	}

	fresh := state.Empty()
	counter := state.Value()
	if fresh {
		counter.Word = input.Word
	}
	counter.Count++
	return nil
}

// [END word_count_mapper]
